sdk/src/test/java/com/horizen/storage/StorageTest.java

```diff
@@ -1,72 +1,159 @@
+            s.update(duplicatedRemoveVersion, new ArrayList<>(), duplicatedRemove);
+            fail("The same value in remove shall thrown exception");
+        }
+        catch (Exception ex) {}
+
+        try {
+            List<Pair<ByteArrayWrapper,ByteArrayWrapper>> duplicatedUpdated = new ArrayList<>(u2);
+            duplicatedUpdated.add(u2.get(0));
+
+            ByteArrayWrapper duplicatedUpdatedVersion = storageFixture.getVersion();
+            s.update(duplicatedUpdatedVersion, duplicatedUpdated, new ArrayList<>());
+            fail("The same value in update shall thrown exception");
+        }
+        catch (Exception ex) {}
+
+        try {
+            s.update(version2, u2, new ArrayList<>());
+            fail("already exist version in update shall thrown exception");
+        }
+        catch (Exception ex) {}
+
+        try {
+            s.rollback(storageFixture.getVersion());
+            fail("Non exist version in rollback shall thrown exception");
+        }
+        catch (Exception ex) {}
+    }
+
+    @Test
+    public void reopenStorage() {
+        File storagePath = storageFixture.tempFile();
+        VersionedLevelDbStorageAdapter s = storageFixture.getStorage(storagePath);
+        assertTrue("Storage expected to be empty.", s.isEmpty());
+
+        assertFalse("Version of empty storage shall be none", s.lastVersionID().isPresent());
+
+        ByteArrayWrapper version0 = storageFixture.getVersion();
+        List<Pair<ByteArrayWrapper,ByteArrayWrapper>> u0 = new ArrayList<>();
+        s.update(version0, u0, new ArrayList<>());
+        List<Pair<ByteArrayWrapper, ByteArrayWrapper>> keysAfterFirstUpdate = s.getAll();
+        assertEquals("No keys are expected after update with empty values", 0, keysAfterFirstUpdate.size());
+        assertEquals("last version shall be as expected after empty update", version0, s.lastVersionID().get());
+
+        int firstUpdateSize = 3;
+        ByteArrayWrapper version1 = storageFixture.getVersion();
+        List<Pair<ByteArrayWrapper,ByteArrayWrapper>> u1 = storageFixture.getKeyValueList(firstUpdateSize);
+        s.update(version1, u1, new ArrayList<>());
+
+        assertFalse("Storage expected to be not empty.", s.isEmpty());
+        assertEquals("Storage must contain 3 items.", firstUpdateSize, s.getAll().size());
+        assertEquals("Storage must contain value for key - " + u1.get(0).getKey(), u1.get(0).getValue(), s.get(u1.get(0).getKey()).get());
+        assertTrue("Storage must contain same elements as sample.", u1.containsAll(s.getAll()));
+        assertEquals("Storage must have specified last version.", version1, s.lastVersionID().get());
+        assertEquals("Storage must have two versions", 2, s.rollbackVersions().size());
+        assertTrue("Storage must have specified versions", s.rollbackVersions().containsAll(Arrays.asList(version0, version1)));
+
+        s.close();
```

Comment:
Probably, we can put re-open test case at the end of the method above, to not to duplicate the code of data population. Not significant.

Reply:
tests shall be independent from each other...